File: backend/app/schemas/analytics.py
from __future__ import annotations

import enum
import uuid
from typing import Optional

from pydantic import BaseModel


class GroupBy(str, enum.Enum):
    day   = "day"
    week  = "week"
    month = "month"


# ─── 1. Appointments by period ────────────────────────────────────────────────

class AppointmentPeriodResponse(BaseModel):
    period: str
    total: int
    cancelled: int
    completed: int


# ─── 2. Top-10 ICD-10 diagnoses ──────────────────────────────────────────────

class TopDiagnosisResponse(BaseModel):
    icd10_code: str
    name_ua: str
    count: int


# ─── 3. Doctor load ───────────────────────────────────────────────────────────

class DoctorLoadResponse(BaseModel):
    doctor_id: uuid.UUID
    full_name: str
    specialization: Optional[str]
    encounters_count: int
    appointments_count: int


# ─── 4. Prescriptions by period ──────────────────────────────────────────────

class PrescriptionPeriodResponse(BaseModel):
    period: str
    total: int
    active: int
    cancelled: int


# ─── 5. Cancellation rate ─────────────────────────────────────────────────────

class CancellationRateResponse(BaseModel):
    total_appointments: int
    cancelled: int
    cancellation_rate: float

File: frontend/src/api/analytics.ts
import client from './client'
import type {
  AppointmentPeriodResponse,
  TopDiagnosisResponse,
  DoctorLoadResponse,
  PrescriptionPeriodResponse,
  CancellationRateResponse,
  GroupBy,
} from './types'

interface DateRange {
  date_from?: string
  date_to?: string
}

export const analyticsApi = {
  appointments: (params: DateRange & { group_by?: GroupBy }) =>
    client.get<AppointmentPeriodResponse[]>('/analytics/appointments', { params }),

  topDiagnoses: (params: DateRange & { doctor_id?: string }) =>
    client.get<TopDiagnosisResponse[]>('/analytics/diagnoses/top10', { params }),

  doctorLoad: (params: DateRange) =>
    client.get<DoctorLoadResponse[]>('/analytics/doctors/load', { params }),

  prescriptions: (params: DateRange & { group_by?: GroupBy }) =>
    client.get<PrescriptionPeriodResponse[]>('/analytics/prescriptions', { params }),

  cancellationRate: (params: DateRange & { doctor_id?: string }) =>
    client.get<CancellationRateResponse>('/analytics/appointments/cancellation-rate', { params }),
}
